### ISSUE TITLE
Attribute and method priority inconsistent between Python and TorchDynamo NNModuleVariable

### ISSUE THREAD

**Issue by yanboliang** [pytorch/torchdynamo#276]
*This is the 6/N issue found from https://github.com/pytorch/torchdynamo/issues/156.*

Minimal code to reproduce:
```
class HeadModule(nn.Module):

    def __init__(self):
        super(HeadModule, self).__init__()

    def forward(self, x):
        return x + 1

class MyModule(nn.Module):

    def __init__(self, **kwargs):
        super(MyModule, self).__init__(**kwargs)
        self.features = None
        self.head = HeadModule()

    def head(self, x):
        return x

    def forward(self, x):
        return self.head(x)

x = torch.ones([2, 2])
model = MyModule()
model.eval()

def my_compiler(gm: torch.fx.GraphModule, example_inputs: List[torch.Tensor]):
    return gm.forward

with torchdynamo.optimize(my_compiler):
    print(model(x))
```
w/o torchdynamo, the output is:
```
tensor([[1., 1.],
        [1., 1.]])
```
w torchdynamo, the output is:
```
tensor([[2., 2.],
        [2., 2.]])
```
The problem is: the attribute ```head``` and the method ```head``` are the same name, Python favors the method, however, torchdynamo prioritizes the attribute. I checked the code in [```NNModuleVariable.var_getattr```](https://github.com/pytorch/torchdynamo/blob/main/torchdynamo/variables/nn_module.py#L108) and found it seems intended behavior. I'd like to ask if this is intended and should we print warning here if we found method of the same name? @jansel

**Comment by jansel** [comment]
This is a bug.  We should change TorchDynamo to ignore `self.__class__.head` if `self.__dict__["head"]` is defined.

**Comment by yanboliang** [comment]
> We should change TorchDynamo to ignore self.__class__.head if self.__dict__["head"] is defined.
I think you mean ignore ```self.__dict__["head"]``` if ```self.__class__.head``` is defined, right?

**Comment by jansel** [comment]
We should match whatever normal python does.

I thought object members took precedence: 
```
>>> class Foo:
...   bar = "on the class"
...   def __init__(self):
...     self.bar = "on the object"
... 
>>> Foo().bar
'on the object'
>>> 
```

Though perhaps I was wrong there.  You should base it on correct python behavior.

**Comment by jansel** [comment]
Trying a bit more with `nn.Modules`:
```
>>> class Bar(torch.nn.Module):
...   def __init__(self):
...     super().__init__()
...     self.hook1 = torch.nn.ReLU()
...     self.hook2 = lambda: "on the object"
...     self.hook3 = "on the object"
...   def hook1(self):
...     return "on the class"
...   def hook2(self):
...     return "on the class"
...   def hook3(self):
...     return "on the class"
... 
>>> Bar().hook1()
'on the class'
>>> Bar().hook2()
'on the object'
>>> Bar().hook3
'on the object'
>>> 
```

As far as I can tell the object one wins, *unless* it is a `torch.nn.Module()` on the class.

I am guessing the same will be true for `torch.Tensor` and `torch.Parameter`.

To replicate the logic we may want to be checking `self.value.__dict__.keys()`:
```
>>> Bar().__dict__.keys()
dict_keys(['training', '_parameters', '_buffers', '_non_persistent_buffers_set', '_backward_hooks', '_is_full_backward_hook', '_forward_hooks', '_forward_pre_hooks', '_state_dict_hooks', '_load_state_dict_pre_hooks', '_modules', 'hook2', 'hook3'])
```

Note hook1 is missing from the keys.

**Comment by yanboliang** [comment]
I did some research and found:
* TLDR: object attribute takes precedence unless it's a ```nn.Module```.

Reason:
* Pytorch [```nn.Module.__setattr__```](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/module.py#L1210) doesn't put ```Module``` and ```Parameter``` in ```object.__dict___```. However, the duplicated class attribute(if there is) is in ```object.__dict___``` and takes precedence.
  ** For ```Parameter```, if there is duplicated class method, it throws error [```attribute xxx already exists```](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/module.py#L353) when initializing the object. So don't need to handle this in torchdynamo side.
* If the attribute is ```torch.Tensor``` type, it fall back to [```object.__setattr__(self, name, value)```](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/module.py#L1255), so will be in ```object.__dict___``` and go through the normal process.

Right now the [fix](https://github.com/pytorch/torchdynamo/pull/278) is to keep behavior consistent with Pytorch, but may be we should enforce Pytorch ```nn.Module``` initialization and check if there is duplicated class attribute for ```Module``` and ```Tensor```, just like what they did for ```Parameter```?

**Comment by jansel** [comment]
I agree.  This might be a good candidate for an upstream change in PyTorch, as this behavior is confusing.  Curious what @albanD thinks.

For now let's make TorchDynamo match the current behavior.

**Comment by albanD** [comment]
In this case, shouldn't Dynamo replicate the `__getattribute__` vs `__getattr__` behavior?
In this case, since the attribute is not in `__dict__`, the `__getattr__` should be called right?

**Comment by jansel** [comment]
Yeah, that is what we are doing.  I was asking if we should change the behavior of `nn.Module` to match python (or throw a warning/error).  The upstream `nn.Module` behavior we are replicating seems wrong.

**Comment by albanD** [comment]
I'm not sure how `nn.Module` doesn't match python?
You mean doesn't match `object`? If so that is to make sure our custom getattr and setattr are always called.

**Comment by jansel** [comment]
```
>>> class Bar(torch.nn.Module):
...   def __init__(self):
...     super().__init__()
...     self.hook1 = torch.nn.ReLU()
...     self.hook2 = lambda: "on the object"
...     self.hook3 = "on the object"
...   def hook1(self):
...     return "on the class"
...   def hook2(self):
...     return "on the class"
...   def hook3(self):
...     return "on the class"
... 
>>> Bar().hook1()
'on the class'
>>> Bar().hook2()
'on the object'
>>> Bar().hook3
'on the object'
>>> 
```

The behavior of hook1 seems confusing/wrong to me.

**Comment by albanD** [comment]
Ho it does lookup on the class object before calling the user-provided `__getattr__`.
We should be able to detect this and raise an error in that case, opened an issue to track that: https://github.com/pytorch/pytorch/issues/78618
